Update parameters.thermal_speed

pheuer opened the pull request:
Scrolling through the issue log, I saw #427 and thought it was an easy addition. However, while making the example I thought that since 1D, 2D, and 3D Maxwellians are included in the distribution.py module, thermal speeds for different dimensionalities should also be supported (if for no other reason than to remind people to use the right one). Does everyone agree this is worthwhile functionality?

In this PR I've added this functionality and wrote appropriate tests. I also created a thermal_speed example notebook that contains the plot originally requested in #427 and have added this example to the documentation.

Closes #427

pep8speaks commented:
Hello @pheuer! Thanks for submitting your pull request.

* In the file [`plasmapy/formulary/parameters.py`](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/parameters.py):

> [Line 425:9](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/parameters.py#L425): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 425:11](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/parameters.py#L425): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

* In the file [`plasmapy/formulary/tests/test_parameters.py`](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/tests/test_parameters.py):

> [Line 1097:1](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/tests/test_parameters.py#L1097): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 1097:21](https://github.com/PlasmaPy/PlasmaPy/blob/7ef6c4e31b091e2eb8cf7477687ed8d35ab42bf7/plasmapy/formulary/tests/test_parameters.py#L1097): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

PlasmaPy follows the [PEP8 style guide](https://www.python.org/dev/peps/pep-0008/), which may be checked locally using [pycodestyle](http://pycodestyle.pycqa.org/en/latest/intro.html)

StanczakDominik commented on docs/formulary/distribution.rst:
```diff
@@ -13,3 +13,4 @@ Distribution functions (`plasmapy.formulary.distribution`)
     :caption: Examples
 
     /notebooks/distribution
+	/notebooks/thermal_speed
```
Looks like a mixed space-tab issue is causing this to throw warnings...
```suggestion
    /notebooks/thermal_speed 
```

StanczakDominik commented on docs/formulary/parameters.rst:
```diff
@@ -11,3 +11,4 @@ Plasma parameters (`plasmapy.formulary.parameters`)
     :caption: Examples
 
     /notebooks/distribution
+	/notebooks/thermal_speed
```
```suggestion
    /notebooks/thermal_speed 
```

StanczakDominik commented on plasmapy/formulary/parameters.py:
```diff
@@ -478,10 +483,34 @@ def thermal_speed(
+    :math:`v_{th}`: most probable, root-mean-square (RMS), or mean magnitude.
+    The value of N in each case is
+
+    .. list-table:: Values of constant N
+       :widths: 50, 25, 25, 25
+       :header-rows: 1
+
+       * - Dim.
+         - Most-Probable
+         - RMS
+         - Mean-Magnitude
+       * - 1D
+         - 0
+         - 1
+         - :math:`2/\pi`
+       * - 2D
+         - 1
+         - 2
+         - :math:`\pi/2`
+       * - 3D
+         - 2
+         - 3
+         - :math:`8/\pi`
```
This is going to be dreadful to read in jupyter/ipython/other shells, but wow, in the rendered docs that is one pretty table. :smile: I'm completely down for that.

pheuer replied:
Yeah - one advantage of calculating these with an explicit dependence on ndim (rather than the if/else tree) would be the ability to more directly state the formulas in the docstring simply for all ndim. 

Side question: how do you get to the docs for a PR (without typing out the link and changing the PR number)?

StanczakDominik replied:
If you look for `docs/readthedocs.org:plasmapy` in the `Checks` tab below, you can click on Details and that should point you to the built docs :)

pheuer replied:
Awesome, thanks!

StanczakDominik commented on plasmapy/formulary/parameters.py:
```diff
@@ -507,15 +536,43 @@ def thermal_speed(
     """
     m = mass if np.isfinite(mass) else particles.particle_mass(particle)
 
+    # TODO: Could implement this more cleanly for any integer ndim
+    # at the cost of using the gamma function in the expression for the
+    # mean magnitude velocity.
+
```
Sure, but it'd be slower, too.

pheuer replied:
Fair enough - 3D plasmas are more than complicated enough for me.

StanczakDominik commented on plasmapy/formulary/parameters.py:
```diff
@@ -507,15 +536,43 @@ def thermal_speed(
-    elif method == "mean_magnitude":
-        V = np.sqrt(8 * k_B * T / (m * np.pi))
+    if ndim == 1:
+        if method == "most_probable":
+            V = 0 * u.m/u.s
+        elif method == "rms":
+            V = np.sqrt(k_B * T / m)
+        elif method == "mean_magnitude":
+            V = np.sqrt(2 * k_B * T / (m * np.pi))
+        else:
+            raise ValueError("Method {method} not supported in thermal_speed")
+
+    elif ndim == 2:
+        if method == "most_probable":
+            V = np.sqrt(k_B * T / m)
+        elif method == "rms":
+            V = np.sqrt(2 * k_B * T / m)
+        elif method == "mean_magnitude":
+            V = np.sqrt(np.pi * k_B * T / (2 * m))
+        else:
+            raise ValueError("Method {method} not supported in thermal_speed")
+
+    elif ndim == 3:
+        if method == "most_probable":
+            V = np.sqrt(2 * k_B * T / m)
+        elif method == "rms":
+            V = np.sqrt(3 * k_B * T / m)
+        elif method == "mean_magnitude":
+            V = np.sqrt(8 * k_B * T / (m * np.pi))
+        else:
+            raise ValueError("Method {method} not supported in thermal_speed")
     else:
-        raise ValueError("Method {method} not supported in thermal_speed")
+        raise ValueError("{ndim} is not a supported value for ndim in "
+                         "thermal_speed")
```
Actually, you know what a nice refactor could be? A dictionary!

```python
coefficients = {1: {"most_probable": 0, ...}, ...}
try:
   coefficient = coefficients[ndim][method]
except KeyError:
    raise ValueError("Method {method} not supported in thermal_speed") 
    # could even break this down into two dict-indexing steps to handle both the exceptions
V = np.sqrt(coefficient * k_B * T / m)
```

The dictionary could even be created outside the function to save some completely unnoticeable amount of time on instantiating it on each call, if we care about that.

pheuer replied:
Well that's a much, much better way to do that! Done.

pheuer commented:
@StanczakDominik Thanks for the suggestions! I've implemented the dictionary refactor: that's very elegant. A few other changes: 

- I ran pre-commit and it worked well (once I updated the fork to the current master: before that it wanted to fix lots of files). 

- I updated the example notebook to have a thumbnail. I ran into the same problem here as I did with the Thomson notebook where multi-figure plots are required but make bad thumbnails, so a single plot must also be included. Maybe theres a better way to handle this through a more sophisticated use of nbsphinx?

StanczakDominik commented:
> I updated the example notebook to have a thumbnail. I ran into the same problem here as I did with the Thomson notebook where multi-figure plots are required but make bad thumbnails, so a single plot must also be included. Maybe theres a better way to handle this through a more sophisticated use of nbsphinx?

I don't think there is, or at least I'm not aware of it yet :(

 I'll give this a quick once-over again and merge it, most likely :)

StanczakDominik commented on plasmapy/formulary/parameters.py:
```diff
@@ -412,6 +412,16 @@ def ion_sound_speed(
+# This dictionary defines coefficients for thermal speeds
+# calculated for different methods and values of ndim.
+# Created here to avoid re-instantiating on each call
+coefficients = {
```
I think I'd like to make these a private variable:
```suggestion
_coefficients = {
```

The underscore is sort of a convention saying "you probably shouldn't need to use this variable outside", since Python lacks `private`s.

StanczakDominik commented on plasmapy/formulary/parameters.py:
```diff
@@ -509,16 +548,16 @@ def thermal_speed(
-        V = np.sqrt(8 * k_B * T / (m * np.pi))
-    else:
+    try:
+        coef = coefficients[ndim]
```
```suggestion
        coef = _coefficients[ndim] 
```

StanczakDominik commented on plasmapy/formulary/tests/test_parameters.py:
```diff
@@ -365,6 +365,44 @@ def test_thermal_speed():
+    # Explicitly check all three modes and dimensionalities
+    # ndim = 1
+    assert np.isclose(thermal_speed(T_e, method="most_probable", ndim=1).si.value, 0.0)
+
```
```suggestion

    # Regression tests start here!
```
I'm bad at keeping track of which of our tests are known value comparisons (probably a minuscule amount) and which are just regression tests, so while I'm looking at this file, lemme add this note. This sort of thing hurt us a bit when astropy changed their current constant values. The implementation is fine, though.

pheuer commented:
@StanczakDominik  Looks good - most of the tests I've written for both this and the Thomson module have been known value comparisons: I can see the potential problems with this approach but I can't think of a better way to ensure the physics equations haven't been modified?

StanczakDominik commented:
> @StanczakDominik  Looks good - most of the tests I've written for both this and the Thomson module have been known value comparisons: I can see the potential problems with this approach but I can't think of a better way to ensure the physics equations haven't been modified?

Me neither, unless the original derivation provided value samples... Which they usually don't. Oh well.

StanczakDominik commented:
We could potentially try property based testing one of these days... I'd have to pick it up from scratch though.

No matter! Merging :)
